### ISSUE TITLE
BUG: np.mean(axis=0) return incorrect value for a large size float32 array

### ISSUE THREAD

**Issue by cesaryuan** [numpy/numpy#22956]
### Describe the issue:

For a large size float32 array with second dimension larger than 1, `np.mean(axis=0)` return incorrect value as shown follow.
```ipython
Python 3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.22.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '1.24.1'

In [3]: np.ones((20000000, 1), dtype=np.float32).mean(axis=0)
Out[3]: array([1.], dtype=float32)

In [4]: np.ones((20000000, 2), dtype=np.float32).mean(axis=0)
Out[4]: array([0.8388608, 0.8388608], dtype=float32)
```

### Reproduce the code example:

```python
import numpy as np
# result: [1.]
print(np.ones((20000000, 1), dtype=np.float32).mean(axis=0))
# result: [0.8388608, 0.8388608]
print(np.ones((20000000, 2), dtype=np.float32).mean(axis=0))
```

### Runtime information:

1.24.1
3.8.8 (default, Apr 13 2021, 15:08:03) [MSC v.1916 64 bit (AMD64)]

**Comment by mahtantaoui** [comment]
Can I please work on this one ??

**Comment by arthurfeeney** [comment]
I thought this would be fun to look into, but now I'm thinking it's just "floating point shenanigans."
It looks like `np.ones((20000000, 1), dtype=np.float32).mean(axis=0)` and `np.ones((20000000, 2), dtype=np.float32).mean()` use a fancy [blocked accumulator](https://github.com/numpy/numpy/blob/307477c454e960e0f6805ead4fd545014efd3701/numpy/core/src/umath/loops_utils.h.src#L78)

while `np.ones((20000000, 2), dtype=np.float32).mean(axis=0)` seems to essentially just do a [simple thing](https://github.com/numpy/numpy/blob/307477c454e960e0f6805ead4fd545014efd3701/numpy/core/src/umath/loops_arithm_fp.dispatch.c.src#L541) where it adds 1 each time.

`20000000` is larger than the single precision float `16777216`, which is like the "max" last integer value you can really accumulate 1 onto, because the next float is actually 16777218, which is 2 away [described here](https://stackoverflow.com/questions/12596695/why-does-a-float-variable-stop-incrementing-at-16777216-in-c). So when the failing case tries to compute the mean, it just adds one each time and gets stuck at `16777216`. 

I think the blocked accumulator adds larger chunks, so it doesn't get stuck trying to increment `16777216`.

Maybe there could be a `stable_mean` or something that always computes the mean in chunks to avoid something like this--IDK, random thought :-)

**Comment by mhvk** [comment]
@arthurfeeney - that seems correct. And we've had other issues noting that it is surprising that if there is a faster-changing axis, the answer can change. If I recall correctly, the difference is that when summing over all but the smallest-stride axis, the iterator processes the data in chunks, in this case making an output array of 2 elements, and then adding all entries along the long axis to it. From a memory access perspective, this is obviously optimal. Not sure what the best solution is, though!

**Comment by RomanKosobrodov** [comment]
I see the same bug when calculating the mean pixel value of a float32 image.

`img = np.ones(shape=(5376, 8368, 3), dtype=np.float32)`
`m = np.mean(w, axis=(0, 1))`

produces this output:
`array([0.37294, 0.37294, 0.37294], dtype=float32)`

Downsampling the image by a factor of 2 
`img = np.ones(shape=(5376//2, 8368//2, 3), dtype=np.float32)`
`m = np.mean(w, axis=(0, 1))`

gives the correct result:
`array([1., 1., 1.], dtype=float32)`

Debugging it was a lot of fun. 

(numpy 1.23.5; python 3.11.0; Windows 10 Home)

**Comment by erdmann** [comment]
I tried to determine whether this was a regression, and it looks like it exists all the way back to at least before numpy=1.7.2 on Python 2.7, which was the earliest version I could check.

Also, FWIW (and as a single-precision workaround), pytorch and jax.numpy both do the calculation correctly in single precision on the CPU.

I'm amazed I never encountered this before in doing image processing, but maybe it's just because the catastrophic loss of precision occurs silently and I just never noticed it.  Having just lost several days to this bug, I suspect it might be a hidden cause of many subtle malfunctions in downstream code.

**Comment by recurrentimh** [comment]
Echoing that this still an issue.

Any package which is supposed to be used for numerical analysis has to be able to compute a mean correctly. 

Until this is fixed I will stop recommending the use of numpy for any commercial or safety critical task.

**Comment by Xin-yang-Liu** [comment]
I got the same issue on the latest release (numpy 2.1.3 with python 3.13). I believe this is a major BUG deserves your attention. @seberg @charris @mattip 
According to arthurfeeney's [comment](https://github.com/numpy/numpy/issues/22956#issuecomment-1374363475).  This seems to be a float number overflow issue. Can we at least raise an error or warning when the float number overflow happens?

**Comment by charris** [comment]
Hmm, recursion with blocks that are powers of 2 might be interesting.

**Comment by charris** [comment]
You keep a local array, say of 64 floats. Add two floats, divide by 2, put the result in the last position. Repeat, but now add the result to the previous, divide by 2, and put in the second to last position. It is like a counter with carry. At the end, put all the results together. For blocks, instead of two, use four, essentially a base four counter. Divisions by powers of two should be efficient, the only messy bit comes when putting the final result together.

**Comment by mattip** [comment]
As @arthurfeeney points out
| 20000000 is larger than the single precision float 16777216, which is like the "max" last integer value you can really accumulate 1 onto, because the next float is actually 16777218, which is 2 away [described here](https://stackoverflow.com/questions/12596695/why-does-a-float-variable-stop-incrementing-at-16777216-in-c). So when the failing case tries to compute the mean, it just adds one each time and gets stuck at 16777216.

i.e.
```
>>> np.float32(16777216) + np.float32(1)
np.float32(16777216.0)
>>> 
```
Using a float64 accumulator would avoid this problem, with or without a change of algorithm.

In the int overflow case we warn
```
>>> np.int8(127) + np.int8(1)
<stdin>:1: RuntimeWarning: overflow encountered in scalar add
np.int8(-128)
```

but there is no such warning here because it is not really "overflow"

**Comment by Xin-yang-Liu** [comment]
@mattip But this float number accuracy issue only happens when you specify axis. If  you calculate mean for the entire array, there is no such issue. Is it possible just to use the same implementation as you did for calculating mean of the entire array?

**Comment by mattip** [comment]
Let's call this `np.float32(16777216) + np.float32(1) == np.float32(16777216)` "max value pinning"

Looking at the example where `a = np.ones((20000000, 2), dtype=np.float32)`.

In the `np.sum(a)` code path, we use `PyUFunc_ReduceWrapper` which [creates a `NpyIter_AdvancedNew`](https://github.com/numpy/numpy/blob/c045306a188f19d7f919512638bd5f02724c9c7e/numpy/_core/src/umath/reduction.c#L283) iterator. This iterator contains a buffer and "chunks" the iterations. On my machine it breaks the loop over `20_000_000 * 2 = 40_000_000` items into chunks of 8192, and uses the `FLOAT_add_AVX2` inner loop to sum each chunk. This chunking is what prevents the max-value-pinning in the example. Knowing the algorithm, we can construct an array that will give very wrong results without the axis argument:
```
>>> a = np.ones((20000000, 2), dtype=np.float32) / 8193;
# prime the first value to be close to the max
>>> a[0,0]= 16777215
>>> print(np.sum(a))
16777216.0
# yup, it does not exceed the max. What should it have been?
>>> print(np.sum(a[1:, :]) + np.sum(a[0, :]))
16782098.0
```

OK, so what happens when using `axis`? Then the iterator works differently: it now iterates across the axis specified, calling the inner loop on the other dimensions. So `np.sum(a, axis=0)` is equivalent to
```
result = np.zeros(2, dtype=a.dtype)
for i in range(a.shape[0]):
    result += a[i, :]
return result
```
Note this calls the inner loop function 20_000_000 times, so it is considerably slower.
Since there is no "chunking" the problem with max-value-pinning is more pronounced.

**Comment by mattip** [comment]
I am not sure what would be faster: using float64 intermediate values or using a better addition algorithm for float32 and float16 like [kahan summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm)

**Comment by mattip** [comment]
Heh, from that article:
| The standard library of the [Python](https://en.wikipedia.org/wiki/Python_(programming_language)) computer language specifies an [fsum](https://docs.python.org/library/math.html#math.fsum) function for accurate summation. Starting with Python 3.12, the built-in "sum()" function uses the Neumaier summation.

**Comment by mattip** [comment]
Here is some code implementing a naive kahan summation. It is not fast but does give the correct answer. To implement something like this in NumPy without changing the NDIter iteration across the axis would mean using some `auxdata`/'userdata' to store the the error correction value `c`. 
<details><summary>kahan.py using cffi</summary>

```
import cffi
import numpy as np
import time

ffi = cffi.FFI()
ffi.cdef("""
    float accurate_sum(float *values, int n);
    float inaccurate_sum(float *values, int n);
""")

ffi.set_source("test_sums", """

static __inline__
void kahan_sum(float *sum, float *c, float y) {
    float t;
    y -= c[0];
    t = sum[0] + y;
    c[0] = (t - sum[0]) - y;
    sum[0] = t;
}

float accurate_sum(float *values, int n) {
    float sum = 0.0;
    float c = 0.0;
    for (int i=0; i<n; i++) {
        kahan_sum(&sum, &c, values[i]);    
    }
    return sum;
}

float inaccurate_sum(float *values, int n){
    float sum = 0.0;
    for (int i=0; i<n; i++)
        sum += values[i];
    return sum;
}
""")

ffi.compile()

from test_sums import ffi, lib

n = 20_000_000
a = np.ones(n, dtype=np.float32)
m = ffi.cast("float*", a.ctypes.data)
start = time.time()
for i in range(10):
    sum = lib.accurate_sum(m, n)
stop = time.time()
print(f"accurate {sum}, took {stop - start:.2f}")
start = time.time()
for i in range(10):
    sum = lib.inaccurate_sum(m, n)
stop = time.time()
print(f"inaccurate {sum}, took {stop - start:.2f}")
```

</details>

On my machine the kahan implementation is 4.9x slower:
```
$ python kahan.py 
accurate 20000000.0, took 0.73
inaccurate 16777216.0, took 0.14
```

**Comment by mark-dostalik** [comment]
I encountered this bug when using `skimage.measure.regionprops` and `intensity_mean` property which uses `np.mean` under the hood. This issue is extremely difficult to debug since no warning is thrown and the loss of precision appears silently.
